File: app/src/main/java/org/asteroidos/sync/connectivity/NotificationService.java
/*
 * AsteroidOSSync
 * Copyright (c) 2024 AsteroidOS
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.asteroidos.sync.connectivity;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

import org.asteroidos.sync.NotificationPreferences;
import org.asteroidos.sync.asteroid.IAsteroidDevice;
import org.asteroidos.sync.dataobjects.Notification;
import org.asteroidos.sync.services.INotificationHandler;
import org.asteroidos.sync.utils.AsteroidUUIDS;

import java.util.HashMap;
import java.util.Objects;
import java.util.UUID;

public class NotificationService implements IConnectivityService, INotificationHandler {

    public static final String TAG = NotificationService.class.toString();
    private final Context mCtx;
    private final IAsteroidDevice mDevice;
    private NotificationReceiver mNReceiver;

    public NotificationService(Context ctx, IAsteroidDevice device) {
        this.mDevice = device;
        this.mCtx = ctx;
    }

    @Override
    public void sync() {
        if (mNReceiver == null) {
            IntentFilter filter = new IntentFilter();
            filter.addAction("org.asteroidos.sync.NOTIFICATION_LISTENER");
            mNReceiver = new NotificationReceiver();
            mCtx.registerReceiver(mNReceiver, filter);

            Intent i = new Intent("org.asteroidos.sync.NOTIFICATION_LISTENER_SERVICE");
            i.putExtra("command", "refresh");
            mCtx.sendBroadcast(i);
        }
    }

    @Override
    public void unsync() {
        if (mNReceiver != null) {
            try {
                mCtx.unregisterReceiver(mNReceiver);
            } catch (IllegalArgumentException ignored) {
            }
            mNReceiver = null;
        }
    }

    @Override
    public final HashMap<UUID, Direction> getCharacteristicUUIDs() {
        HashMap<UUID, Direction> chars = new HashMap<>();
        chars.put(AsteroidUUIDS.NOTIFICATION_UPDATE_CHAR, Direction.TO_WATCH);
        chars.put(AsteroidUUIDS.NOTIFICATION_FEEDBACK_CHAR, Direction.FROM_WATCH);
        return chars;
    }

    @Override
    public final UUID getServiceUUID() {
        return AsteroidUUIDS.NOTIFICATION_SERVICE_UUID;
    }

    @Override
    public void postNotification(Context context, Intent intent) {
        String event = intent.getStringExtra("event");
        if (Objects.equals(event, "posted")) {
            String packageName = intent.getStringExtra("packageName");
            NotificationPreferences.putPackageToSeen(context, packageName);
            NotificationPreferences.NotificationOption notificationOption =
                    NotificationPreferences.getNotificationPreferenceForApp(context, packageName);
            if (notificationOption == NotificationPreferences.NotificationOption.NO_NOTIFICATIONS)
                return;

            int id = intent.getIntExtra("id", 0);
            String appName = intent.getStringExtra("appName");
            String appIcon = intent.getStringExtra("appIcon");
            String summary = intent.getStringExtra("summary");
            String body = intent.getStringExtra("body");
            String vibration;
            if (notificationOption == NotificationPreferences.NotificationOption.SILENT_NOTIFICATION)
                vibration = "none";
            else if (notificationOption == null
                    || notificationOption == NotificationPreferences.NotificationOption.NORMAL_VIBRATION
                    || notificationOption == NotificationPreferences.NotificationOption.DEFAULT)
                vibration = "normal";
            else if (notificationOption == NotificationPreferences.NotificationOption.STRONG_VIBRATION)
                vibration = "strong";
            else if(notificationOption == NotificationPreferences.NotificationOption.RINGTONE_VIBRATION)
                vibration = "ringtone";
            else
                throw new IllegalArgumentException("Not all options handled");

            if(intent.hasExtra("vibration"))
                vibration = intent.getStringExtra("vibration");

            Notification notification = new Notification(
                    Notification.MsgType.POSTED,
                    packageName,
                    id,
                    appName,
                    appIcon,
                    summary,
                    body,
                    vibration);

            mDevice.send(AsteroidUUIDS.NOTIFICATION_UPDATE_CHAR, notification.toBytes(), NotificationService.this);
        } else if (Objects.equals(event, "removed")) {
            int id = intent.getIntExtra("id", 0);

            mDevice.send(AsteroidUUIDS.NOTIFICATION_UPDATE_CHAR, new Notification(Notification.MsgType.REMOVED, id).toBytes(), NotificationService.this);
        }
    }

    class NotificationReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            postNotification(context, intent);
        }
    }
}

File: app/src/main/java/org/asteroidos/sync/dbus/NotificationService.kt
/*
 * AsteroidOSSync
 * Copyright (c) 2024 AsteroidOS
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.asteroidos.sync.dbus

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.util.Log
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import org.asteroidos.sync.NotificationPreferences
import org.asteroidos.sync.NotificationPreferences.NotificationOption
import org.asteroidos.sync.services.INotificationHandler
import org.freedesktop.Notifications
import org.freedesktop.Notifications.NotificationClosed
import org.freedesktop.dbus.connections.impl.DBusConnection
import org.freedesktop.dbus.interfaces.DBusSigHandler
import org.freedesktop.dbus.types.UInt32
import org.freedesktop.dbus.types.Variant
import java.util.Arrays
import java.util.Map
import java.util.Objects

class NotificationService(private val mCtx: Context, private val connectionProvider: IDBusConnectionProvider) : INotificationHandler, DBusSigHandler<NotificationClosed?> {
    private val mapping: BiMap<String, UInt32> = HashBiMap.create()
    private var mNReceiver: NotificationReceiver? = null

    override fun postNotification(context: Context, intent: Intent) {
        val event = intent.getStringExtra("event")
        if (event == "posted") {
            val packageName = intent.getStringExtra("packageName")
            NotificationPreferences.putPackageToSeen(context, packageName)
            val notificationOption = NotificationPreferences.getNotificationPreferenceForApp(context, packageName)
            if (notificationOption == NotificationOption.NO_NOTIFICATIONS) return
            val key = intent.getStringExtra("key")!!
            val appName = intent.getStringExtra("appName")
            val appIcon = intent.getStringExtra("appIcon")
            val summary = intent.getStringExtra("summary")
            val body = intent.getStringExtra("body")
            val vibration: String
            vibration = if (notificationOption == NotificationOption.SILENT_NOTIFICATION) "notif_silent" else if (notificationOption == null || notificationOption == NotificationOption.NORMAL_VIBRATION || notificationOption == NotificationOption.DEFAULT) "notif_normal" else if (notificationOption == NotificationOption.STRONG_VIBRATION) "notif_strong" else if (notificationOption == NotificationOption.RINGTONE_VIBRATION) "ringtone" else throw IllegalArgumentException("Not all options handled")
            connectionProvider.acquireDBusConnection { notify: DBusConnection ->
                synchronized(mapping) {
                    mapping[key] = notify.getRemoteObject("org.freedesktop.Notifications", "/org/freedesktop/Notifications", Notifications::class.java)
                            .Notify(appName, mapping.getOrDefault(key, UInt32(0)), appIcon, summary, body, emptyList(),
                                    Map.of<String, Variant<*>>(
                                            "x-nemo-feedback", Variant(vibration),
                                            "x-nemo-preview-body", Variant(body),
                                            "x-nemo-preview-summary", Variant(summary),
                                            "urgency", Variant(3.toByte())), 0)
                }
            }
        } else if (event == "removed") {
            val key = Objects.requireNonNull(intent.getStringExtra("key"))
            // Avoid an infinite loop when the user dismisses the notification on the watch
            if (mapping.containsKey(key)) {
                var id: UInt32?
                synchronized(mapping) {
                    id = mapping[key]
                    mapping.remove(key)
                }
                connectionProvider.acquireDBusConnection { notify: DBusConnection ->
                    notify.getRemoteObject("org.freedesktop.Notifications", "/org/freedesktop/Notifications", Notifications::class.java)
                            .CloseNotification(id)
                }
            }
        }
    }

    override fun handle(s: NotificationClosed?) {
        synchronized(mapping) {
            if (s != null
                    && s.reason.toInt() == 2
                    && mapping.containsValue(s.id)) {
                val dismiss = Intent("org.asteroidos.sync.NOTIFICATION_LISTENER_SERVICE")
                dismiss.putExtra("command", "dismiss")
                dismiss.putExtra("key", mapping.inverse()[s.id])
                mapping.inverse().remove(s.id)
                mCtx.sendBroadcast(dismiss)
            }
        }
    }

    override fun sync() {
        if (mNReceiver == null) {
            val filter = IntentFilter()
            filter.addAction("org.asteroidos.sync.NOTIFICATION_LISTENER")
            mNReceiver = NotificationReceiver()
            mCtx.registerReceiver(mNReceiver, filter)
            val i = Intent("org.asteroidos.sync.NOTIFICATION_LISTENER_SERVICE")
            i.putExtra("command", "refresh")
            mCtx.sendBroadcast(i)
        }
        connectionProvider.acquireDBusConnection { notify: DBusConnection -> notify.addSigHandler(NotificationClosed::class.java, notify.getRemoteObject("org.freedesktop.Notifications", "/org/freedesktop/Notifications", Notifications::class.java), this@NotificationService) }
    }

    override fun unsync() {
        connectionProvider.acquireDBusConnection { notify: DBusConnection -> notify.removeSigHandler(NotificationClosed::class.java, notify.getRemoteObject("org.freedesktop.Notifications", "/org/freedesktop/Notifications", Notifications::class.java), this@NotificationService) }
        if (mNReceiver != null) {
            try {
                mCtx.unregisterReceiver(mNReceiver)
            } catch (ignored: IllegalArgumentException) {
            }
            mNReceiver = null
        }
    }

    internal inner class NotificationReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            postNotification(context, intent)
        }
    }
}